itinerant-iterators/server/find_bounds.py

```diff
@@ -0,0 +1,127 @@
+        regex (bool, optional): If the text is regex. Defaults to False.
+
+    Returns:
+        list[tuple[int, int, int, int]]: List of bounding boxes for each character to hide
+    """
+    print("Creating a named temporary file..")
+    temp = tempfile.NamedTemporaryFile()
+    os.environ['GOOGLE_APPLICATION_CREDENTIALS'] = temp.name
+
+    temp.write(bytes(json.dumps(dict(os.environ)), 'utf-8'))
+    temp.seek(0)
+    client = vision.ImageAnnotatorClient()
+    temp.close()
+    bounds = []
+
+    image = vision.Image(content=image_content)
+
+    response = client.document_text_detection(image=image)
+    document = response.full_text_annotation
+    if (text.lower() not in document.text.lower() and not regex):
+        return []
+
+    if regex:
+        matches = re.findall(text, document.text)
+        if len(matches) > 0:
+            if isinstance(matches[0], tuple):
+                text = list(itertools.chain(*matches))
+            else:
+                text = matches
+    image_chars = []
+    for page in document.pages:
+        for block in page.blocks:
+            for paragraph in block.paragraphs:
+                for word in paragraph.words:
+                    for symbol in word.symbols:
+                        if not regex:
+                            image_chars.append([
+                                symbol.text.lower(),
+                                (symbol.bounding_box.vertices[0].x,
+                                 symbol.bounding_box.vertices[0].y,
+                                 symbol.bounding_box.vertices[2].x,
+                                 symbol.bounding_box.vertices[2].y)
+                            ])
+                        else:
+                            image_chars.append([
+                                symbol.text,
+                                (symbol.bounding_box.vertices[0].x,
+                                 symbol.bounding_box.vertices[0].y,
+                                 symbol.bounding_box.vertices[2].x,
+                                 symbol.bounding_box.vertices[2].y)
+                            ])
+    if not regex:
+        text = text.lower()
+        text_index = 0
+        temp_bounds = []
+        for image_char in image_chars:
+            try:
+                while text[text_index] == " ":
+                    text_index += 1
+            except IndexError as e:
+                print(e)
+                text_index = 0
+                temp_bounds.clear()
+            if text[text_index].lower() == image_char[0].lower():
+                text_index += 1
+                temp_bounds.append(image_char[1])
+                if text_index == len(text):
+                    bounds.extend(temp_bounds.copy())
+                    temp_bounds.clear()
+                    text_index = 0
+            elif text[text_index].lower() != image_char[0].lower():
+                text_index = 0
+                temp_bounds.clear()
+    else:
```

Comment:
This is frankly hard to follow. I would suggest adding comments and changing variable names to be more descriptive. With simple logic, it's possible to get away with bad variable names, but once there's some complexity, good names and comments are crucial!

> From what I can glean, this seems to go through each character in the image (as detected by GCV) and checks whether it matches a character in the text to filter. If there's a match, then the next character in the image is checked to see whether it matches the next character of the filtered text (while skipping spaces) and so on...
>
> This is a reasonably efficient and straightforward algorithm! (despite the readability issues with the implementation)

(A good start would be including the description I just provided or your own in a comment as a heads-up/aid to  the reader.)

For example, this is a rewrite of this code with better names and with a different approach to keeping track of what characters of the text we're looking for have been found:

```python
    if not regex:
        search_characters = list(text.lower().replace(" ", ""))
        remaining_search_characters = search_characters.copy()
        current_matching_bounds = []
        for image_char, image_char_bound in image_chars:
            # Grab the next character we're still looking for.
            try:
                search_char = remaining_search_characters.pop(0)
            except IndexError:
                # No more characters are left to check. Restart.
                remaining_search_characters = search_characters.copy()
                current_matching_bounds.clear()

            if search_char == image_char.lower():
                current_matching_bounds.append(image_char_bound)
                if len(remaining_search_characters) == 0:
                    bounds.extend(current_matching_bounds)
                    remaining_search_characters = search_characters.copy()
                    current_matching_bounds.clear()
            else:
                # A character did not match, bail and reset the search attempt
                remaining_search_characters = search_characters.copy()
                current_matching_bounds.clear()
```

I did not test this, so this may contain bugs and may not produce the exact same results, but the idea behind my feedback is there.